srv/salt/_modules/osd.py

```diff
@@ -1805,6 +1869,159 @@ def _prefer_underscores(self, devicenames):
+    """
+
+    def __init__(self):
+        """
+        Set the default filename.
+        """
+        self.filename = "/etc/ceph/destroyedOSDs.yml"
+
+        # Keep yaml human readable/editable
+        self.friendly_dumper = yaml.SafeDumper
+        self.friendly_dumper.ignore_aliases = lambda self, data: True
+
+    def update(self, device, osd_id, force=False):
+        """
+        Add the by-path version of device and osd_id.  If by-path does not
+        exist, record current device and issue exception with instructions.
+        If forced, record current device.
+        """
+        content = {}
+        if os.path.exists(self.filename):
+            with open(self.filename, 'r') as destroyed:
+                content = yaml.safe_load(destroyed)
+                log.debug("content: {} {}".format(type(content), content))
+
+        if device in content:
+            # Exit early, no by-pass equivalent from previous run
+            return ""
+
+        by_path = self._by_path(device)
+        if by_path and not force:
+            content[by_path] = osd_id
+        else:
+            # by-path device is missing, save current device to allow
+            # the OSD to be removed and rely on admin following instructions
+            # below OR admin is overriding the save manually with the new
+            # device name.  In either case, save the device name with the ID.
+            content[device] = osd_id
+
+        with open(self.filename, 'w') as destroyed:
+            destroyed.write(yaml.dump(content, Dumper=self.friendly_dumper,
+                            default_flow_style=False))
+
+        if by_path or force:
+            return ""
+
+        # Hard enough to read without the else indent
+        example = '/dev/disk/by-id/scsi-SATA_INTEL_SSDSC2BA40_BTHV6082036Q400NGN'
+        msg = ("Device {} is missing a /dev/disk/by-path symlink.\n"
+               "Device cannot be replaced automatically.\n\n"
+               "Replace the device, find the new device name and run\n\n"
+               "salt {} osd.update_destroyed {} force=True"
+               ).format(device, __grains__['id'], example)
+        log.error(msg)
+        return msg
+
+    def get(self, device):
+        """
+        Return ID
+        """
+        if os.path.exists(self.filename):
+            with open(self.filename, 'r') as destroyed:
+                content = yaml.safe_load(destroyed)
+            by_path = self._by_path(device)
+            if by_path and by_path in content:
+                return content[by_path]
+        return ""
+
+    # pylint: disable=no-self-use
+    def _by_path(self, device):
+        """
+        Return the equivalent by-path device name
+        """
+        cmd = (r"find -L /dev/disk/by-path -samefile {}".format(device))
+        _rc, _stdout, _stderr = __salt__['helper.run'](cmd)
+        if _stdout:
+            _devices = _stdout.split()
+            if _devices[0]:
```

Comment:
Even the conditional check on a potentially empty array will raise an "IndexError":

It'd rather:
``` python
if devices:
  return devices[0]
```

Reply:
This could still produce an IndexError.
```if devices and devices[0]``` should do the trick. What are the repercussions though if we don't throw something here? Might be a good idea to fail loudly here?

Reply:
> This could still produce an IndexError.

The `if devices` only returns True if there is at least one item in the list. And that's what we are asking for here - the first element. Since python does not allow to set custom indices ( or at least we don't do that here ) we are fine with that check, aren't we?

``` python
In [1]: foo = ['bar']

In [2]: if foo:
   ...:     print(foo[0])
   ...: 
bar

In [3]: foo = []

In [4]: if foo:
   ...:     print(foo[0])
   ...: 

In [5]: 
```

Reply:
Yes you're right...we can rely on having a list here.

Reply:
changed